mesonbuild/ast/interpreter.py

```diff
@@ -99,13 +100,72 @@ class IntrospectionBuildTarget(MesonInterpreterObject):
     kwargs: T.Dict[str, TYPE_var]
     node: FunctionNode
 
+class DataflowDAG:
+    src_to_tgts: T.DefaultDict[T.Union[BaseNode, UnknownValue], T.Set[T.Union[BaseNode, UnknownValue]]]
+    tgt_to_srcs: T.DefaultDict[T.Union[BaseNode, UnknownValue], T.Set[T.Union[BaseNode, UnknownValue]]]
+
+    def __init__(self) -> None:
+        self.src_to_tgts = defaultdict(set)
+        self.tgt_to_srcs = defaultdict(set)
+
+    def add_edge(self, source: T.Union[BaseNode, UnknownValue], target: T.Union[BaseNode, UnknownValue]) -> None:
+        self.src_to_tgts[source].add(target)
+        self.tgt_to_srcs[target].add(source)
+
+    # Returns all nodes in the DAG that are reachable from a node in `srcs`.
+    # In other words, A node `a` is part of the returned set exactly if data
+    # from `srcs` flows into `a`, directly or indirectly.
+    # Certain edges are ignored.
+    def reachable(self, srcs: T.Set[T.Union[BaseNode, UnknownValue]], reverse: bool) -> T.Set[T.Union[BaseNode, UnknownValue]]:
+        reachable = srcs.copy()
+        active = srcs.copy()
+        while active:
+            new: T.Set[T.Union[BaseNode, UnknownValue]] = set()
+            if reverse:
+                for tgt in active:
+                    new.update(src for src in self.tgt_to_srcs[tgt] if not ((isinstance(src, FunctionNode) and src.func_name.value not in {'files', 'get_variable'}) or isinstance(src, MethodNode)))
```

Comment:
The condition is used three times, and is rather complex. Maybe should you extract it as a function?

Reply:
Done